File: app/src/main/java/org/ole/planet/myplanet/service/AutoSyncWorker.java
package org.ole.planet.myplanet.service;

import static android.content.Context.MODE_PRIVATE;
import static org.ole.planet.myplanet.ui.dashboard.DashboardActivity.MESSAGE_PROGRESS;
import static org.ole.planet.myplanet.ui.sync.SyncActivity.PREFS_NAME;

import android.app.ActivityManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import org.ole.planet.myplanet.MainApplication;
import org.ole.planet.myplanet.callback.SuccessListener;
import org.ole.planet.myplanet.callback.SyncListener;
import org.ole.planet.myplanet.datamanager.Service;
import org.ole.planet.myplanet.model.Download;
import org.ole.planet.myplanet.model.MyPlanet;
import org.ole.planet.myplanet.ui.sync.LoginActivity;
import org.ole.planet.myplanet.utilities.Constants;
import org.ole.planet.myplanet.utilities.DialogUtils;
import org.ole.planet.myplanet.utilities.FileUtils;
import org.ole.planet.myplanet.utilities.Utilities;

import java.util.Date;

public class AutoSyncWorker extends Worker implements SyncListener, Service.CheckVersionCallback, SuccessListener {
    private SharedPreferences preferences;
    private Context context;

    public AutoSyncWorker(
            @NonNull Context context,
            @NonNull WorkerParameters workerParams
    ) {
        super(context, workerParams);
        this.context = context;
    }

    @NonNull
    @Override
    public Result doWork() {
        preferences = context.getSharedPreferences(SyncManager.PREFS_NAME, MODE_PRIVATE);
        long lastSync = preferences.getLong("LastSync", 0);
        long currentTime = System.currentTimeMillis();
        int syncInterval = preferences.getInt("autoSyncInterval", 60 * 60);

        if ((currentTime - lastSync) > (syncInterval * 1000)) {
            // Post a Runnable to the main thread's Handler to show the Toast
            Handler mainHandler = new Handler(Looper.getMainLooper());
            mainHandler.post(() -> {
                Utilities.toast(context, "Syncing started...");
            });
            new Service(context).checkVersion(this, preferences);
        }

        return Result.success();
    }

    private boolean isMyServiceRunning(Class<?> serviceClass) {
        ActivityManager manager = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
        for (ActivityManager.RunningServiceInfo service : manager.getRunningServices(Integer.MAX_VALUE)) {
            if (serviceClass.getName().equals(service.service.getClassName())) {
                return true;
            }
        }
        return false;
    }

    public BroadcastReceiver broadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (intent.getAction().equals(MESSAGE_PROGRESS)) {
                Download download = intent.getParcelableExtra("download");
                if (!download.getFailed() && download.getCompleteAll()) {
                    FileUtils.installApk(context, download.getFileUrl());
                }
            }
        }
    };

    @Override
    public void onSyncStarted() {
        Utilities.log("Sync started " + new Date());
    }

    @Override
    public void onSyncComplete() {
        Utilities.log("Sync completed");
    }

    @Override
    public void onSyncFailed(String msg) {
        if (MainApplication.syncFailedCount > 3) {
            context.startActivity(new Intent(context, LoginActivity.class).putExtra("showWifiDialog", true).setFlags(Intent.FLAG_ACTIVITY_NEW_TASK));
        }
    }

    @Override
    public void onUpdateAvailable(MyPlanet info, boolean cancelable) {
        if (Constants.showBetaFeature(Constants.KEY_AUTOUPDATE, context)) {
            DialogUtils.startDownloadUpdate(context, Utilities.getApkUpdateUrl(info.getLocalapkpath()), null);
        }
    }

    @Override
    public void onCheckingVersion() {

    }

    @Override
    public void onError(String msg, boolean blockSync) {
        if (!blockSync) {
            SyncManager.getInstance().start(this);
            UploadToShelfService.getInstance().uploadUserData(success -> new Service(MainApplication.context).healthAccess(success1 -> UploadToShelfService.getInstance().uploadHealth()));
            if (!MainApplication.isSyncRunning) {
                MainApplication.isSyncRunning = true;
                UploadManager.getInstance().uploadExamResult(this);
                UploadManager.getInstance().uploadFeedback(this);
                UploadManager.getInstance().uploadAchievement();
                UploadManager.getInstance().uploadResourceActivities("");
                UploadManager.getInstance().uploadUserActivities(this);
                UploadManager.getInstance().uploadCourseActivities();
                UploadManager.getInstance().uploadSearchActivity();
                UploadManager.getInstance().uploadRating(this);
                UploadManager.getInstance().uploadResource(this);
                UploadManager.getInstance().uploadNews();
                UploadManager.getInstance().uploadTeams();
                UploadManager.getInstance().uploadTeamTask();
                UploadManager.getInstance().uploadCrashLog(this);
                UploadManager.getInstance().uploadActivities(success -> MainApplication.isSyncRunning = false);
            }
        }
    }

    @Override
    public void onSuccess(String s) {
        SharedPreferences settings = MainApplication.context.getSharedPreferences(PREFS_NAME, MODE_PRIVATE);
        settings.edit().putLong("lastUsageUploaded", new Date().getTime()).commit();

    }
}

File: app/src/main/java/org/ole/planet/myplanet/service/AutoSyncWorker.kt
package org.ole.planet.myplanet.service

import android.app.ActivityManager
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.util.Log
import androidx.core.content.edit
import androidx.work.Worker
import androidx.work.WorkerParameters
import dagger.hilt.android.EntryPointAccessors
import java.util.Date
import java.util.concurrent.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import org.ole.planet.myplanet.MainApplication
import org.ole.planet.myplanet.callback.OnSuccessListener
import org.ole.planet.myplanet.callback.OnSyncListener
import org.ole.planet.myplanet.data.DataService
import org.ole.planet.myplanet.data.DataService.CheckVersionCallback
import org.ole.planet.myplanet.di.AutoSyncEntryPoint
import org.ole.planet.myplanet.model.MyPlanet
import org.ole.planet.myplanet.service.sync.SyncManager
import org.ole.planet.myplanet.ui.sync.LoginActivity
import org.ole.planet.myplanet.utilities.Constants.PREFS_NAME
import org.ole.planet.myplanet.utilities.DialogUtils.startDownloadUpdate
import org.ole.planet.myplanet.utilities.UrlUtils
import org.ole.planet.myplanet.utilities.Utilities

class AutoSyncWorker(
    private val context: Context,
    workerParams: WorkerParameters
) : Worker(context, workerParams), OnSyncListener, CheckVersionCallback, OnSuccessListener {
    private lateinit var preferences: SharedPreferences
    private lateinit var syncManager: SyncManager
    private lateinit var uploadManager: UploadManager
    private lateinit var uploadToShelfService: UploadToShelfService
    private val workerScope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    override fun doWork(): Result {
        if (isStopped) return Result.success()
        val entryPoint = EntryPointAccessors.fromApplication(context, AutoSyncEntryPoint::class.java)
        preferences = entryPoint.sharedPreferences()
        syncManager = entryPoint.syncManager()
        uploadManager = entryPoint.uploadManager()
        uploadToShelfService = entryPoint.uploadToShelfService()
        val lastSync = preferences.getLong("LastSync", 0)
        val currentTime = System.currentTimeMillis()
        val syncInterval = preferences.getInt("autoSyncInterval", 60 * 60)
        if (currentTime - lastSync > syncInterval * 1000) {
            if (isAppInForeground(context)) {
                Utilities.toast(context, "Syncing started...")
            }
            DataService(context).checkVersion(this, preferences)
        }
        return Result.success()
    }

    override fun onSyncStarted() {}

    override fun onSyncComplete() {}

    override fun onSyncFailed(msg: String?) {
        if (MainApplication.syncFailedCount > 3) {
            context.startActivity(Intent(context, LoginActivity::class.java)
                .putExtra("showWifiDialog", true)
                .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK))
        }
    }

    override fun onUpdateAvailable(info: MyPlanet?, cancelable: Boolean) {
        startDownloadUpdate(context, UrlUtils.getApkUpdateUrl(info?.localapkpath), null, workerScope)
    }

    override fun onCheckingVersion() {}
    override fun onError(msg: String, blockSync: Boolean) {
        if (!blockSync) {
            syncManager.start(this, "upload")
            uploadToShelfService.uploadUserData {
                DataService(MainApplication.context).healthAccess {
                    uploadToShelfService.uploadHealth()
                }
            }
            if (!MainApplication.isSyncRunning) {
                MainApplication.isSyncRunning = true
                workerScope.launch {
                    try {
                        uploadManager.uploadExamResult(this@AutoSyncWorker)
                        uploadManager.uploadFeedback()
                        uploadManager.uploadAchievement()
                        uploadManager.uploadResourceActivities("")
                        uploadManager.uploadUserActivities(this@AutoSyncWorker)
                        uploadManager.uploadCourseActivities()
                        uploadManager.uploadSearchActivity()
                        uploadManager.uploadRating()
                        uploadManager.uploadResource(this@AutoSyncWorker)
                        uploadManager.uploadNews()
                        uploadManager.uploadTeams()
                        uploadManager.uploadTeamTask()
                        uploadManager.uploadMeetups()
                        uploadManager.uploadAdoptedSurveys()
                        uploadManager.uploadCrashLog()
                        uploadManager.uploadSubmissions()
                        uploadManager.uploadActivities(null)
                    } catch (e: CancellationException) {
                        // Worker was stopped - this is expected, don't treat as failure
                        Log.d("AutoSyncWorker", "Upload cancelled - worker stopped")
                    } catch (e: Exception) {
                        Log.e("AutoSyncWorker", "error: ${e.message}")
                        onSyncFailed(e.message)
                    } finally {
                        MainApplication.isSyncRunning = false
                    }
                }
            }
        }
    }

    override fun onSuccess(success: String?) {
        val settings = MainApplication.context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        settings.edit { putLong("lastUsageUploaded", Date().time) }
    }

    override fun onStopped() {
        super.onStopped()
        workerScope.cancel()
        MainApplication.isSyncRunning = false
    }

    private fun isAppInForeground(context: Context): Boolean {
        val activityManager = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        val runningProcesses = activityManager.runningAppProcesses ?: return false

        for (processInfo in runningProcesses) {
            if (processInfo.processName == context.packageName &&
                processInfo.importance == ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND) {
                return true
            }
        }
        return false
    }
}
